feat: allow adding series

joaoviana opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

Closes: #10863 

### Description:

- Can add/edit/remove to the y axis
- Can edit its aggregation
- When there's more than one, it turns into a show/hide stack of bar series
- doesn't work with other types **yet**
- Aggregation config changes from a segmented control to a better styled Select
- 


DEMO:


https://github.com/user-attachments/assets/fd984b33-f75a-48a7-a06d-8e2d66371cfb



> [!NOTE]
> I want to split the components in the configuration component a bit better - but shall be done when we prioritise more the UI refactoring.

> [!NOTE]
> I can't yet change the colour and label of the series - I want to do this on a separate PR (https://github.com/lightdash/lightdash/issues/10947) after this PR's review

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

joaoviana commented on packages/frontend/src/features/sqlRunner/store/barChartSlice.ts:
```diff
@@ -175,6 +163,32 @@ export const barChartConfigSlice = createSlice({
                 config.display.yAxis[index].position = position;
             }
         },
+        addSeries: (
+            state,
+            action: PayloadAction<
+                SqlTransformBarChartConfig['y'][number]['reference']
+            >,
+        ) => {
+            if (!state.config) return;
+            if (!state.config.fieldConfig) return;
+
+            const yAxisFields = state.config.fieldConfig.y;
+            if (yAxisFields) {
+                state.config.fieldConfig.y.push({
+                    reference: action.payload,
+                    aggregation: DEFAULT_AGGREGATION,
+                });
+                state.config.display = state.config.display || {};
+                state.config.display.yAxis = state.config.display.yAxis || [];
+                state.config.display.series = {
+                    ...state.config.display.series,
+                    [action.payload]: {
+                        yAxisIndex: yAxisFields.length,
+                        label: action.payload,
+                    },
+                };
+            }
+        },
```
is this what is expected?
When I add a series I also push this new item to the `yAxis`? 🤔

joaoviana commented on packages/frontend/src/components/VisualizationConfigs/common/Config.tsx:
```diff
@@ -20,6 +21,12 @@ const Heading: FC<PropsWithChildren> = ({ children }) => (
     </Text>
 );
 
+const Subheading: FC<PropsWithChildren> = ({ children }) => (
```
I had this a long time ago but removed because wasn't using it - finally found a place where this is necessary!

magnew replied:
🙌

joaoviana commented on packages/frontend/src/features/sqlRunner/components/BarChartAggregationConfig.tsx:
```diff
@@ -58,18 +75,35 @@ export const BarChartAggregationConfig: FC<Props> = ({
             onChange={(value) => value && onChangeAggregation(value)}
+            styles={(theme) => ({
+                input: {
+                    width: '110px',
```
can't use any of mantine's size keywords - this will do

magnew replied:
As soon as you do something a tiny bit custom...

joaoviana commented on packages/frontend/src/features/sqlRunner/components/BarChartFieldConfiguration.tsx:
```diff
@@ -234,15 +286,21 @@ export const BarChartFieldConfiguration = () => {
+                        yAxisFields.map((field, index) => (
                             <YFieldsAxisConfig
-                                key={field.reference}
+                                key={field.reference + index}
```
I think this is safe - or we have IDs generated on the fly

magnew replied:
This is fine

joaoviana commented on packages/frontend/src/features/sqlRunner/store/barChartSlice.ts:
```diff
@@ -78,14 +66,12 @@ export const barChartConfigSlice = createSlice({
-                    (axis) =>
-                        axis.reference === action.payload.previousReference,
-                );
+                const yAxis = config.fieldConfig.y[action.payload.index];
```
fixed an issue with the y axis not being updated correctly (now that we allow multiple, it should be handled this way)

owlas replied:
@joaoviana you can remove the `previousReference` I think this was only being used because we weren't passing `index`

joaoviana replied:
good spot, done here: https://github.com/lightdash/lightdash/pull/10932/commits/84e5cbab2a5a6ed2a9c5522359ded287278c79b5

magnew reviewed (commented):
This looks amazing -- fun to use!

There's a little issue with removing series where the series goes away, but there is still space for it and it's still in the legend. Here I added and removed two:

<img width="1616" alt="Screenshot 2024-08-01 at 09 28 46" src="https://github.com/user-attachments/assets/c54e94fe-bcba-4c28-8933-8a54c9185bf6">

joaoviana commented:
> This looks amazing -- fun to use!
> 
> There's a little issue with removing series where the series goes away, but there is still space for it and it's still in the legend. Here I added and removed two:
> 
> <img alt="Screenshot 2024-08-01 at 09 28 46" width="1616" src="https://private-user-images.githubusercontent.com/1864179/354124333-c54e94fe-bcba-4c28-8933-8a54c9185bf6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1MDI3MzUsIm5iZiI6MTcyMjUwMjQzNSwicGF0aCI6Ii8xODY0MTc5LzM1NDEyNDMzMy1jNTRlOTRmZS1iY2JhLTRjMjgtODkzMy04YTU0YzkxODViZjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDFUMDg1MzU1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTRmZTBmOTIyM2E0YjFiNWQ1OGNmYzMxY2NmYzhjNTNhNDU4YTQ0NDAwNThjMDJiN2FlMmIxZGZlNTBhM2VhNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.AWWaRzW4-K_LOze_5YY7IrjewV-C3y46ubh9EpriFfw">

Thanks for sharing this @magnew  - I tried to replicate and couldn't - I believe this has been fixed with the introduction of `notMerge` - But will retest on render now that it's rebased 👍

owlas commented on packages/frontend/src/features/sqlRunner/store/barChartSlice.ts:
```diff
@@ -178,6 +164,38 @@ export const barChartConfigSlice = createSlice({
+            const yAxisFieldsAvailable = state.options.yLayoutOptions.filter(
+                (option) =>
+                    option.reference !==
+                    state.config?.fieldConfig?.y[0]?.reference,
```
Shouldn't this be checking all the y fields not only `y[0]`

joaoviana replied:
Not too sure 🤷 I was thinking more like: choose whatever is not first - but I can update to check for all that are active!

joaoviana replied:
done and tested here: https://github.com/lightdash/lightdash/pull/10932/commits/2c0cb4807e8d50661bd103d26c0af93eab87c951
